Source/Applications/SystemCenterNotification/Model/UserInformation.cs

```diff
@@ -204,5 +224,96 @@ public IHttpActionResult PatchPhone(string phone)
                 return InternalServerError(ex);
             }
         }
+
+        private bool IsValidADUser(string userName)
+        {
+            string sid = UserInfo.GroupNameToSID(userName);
+            return UserInfo.IsUserSID(sid);
+        }
+
+        private bool IsValidAzureADUserName(string userName)
+        {
+            if (userName.Contains("\\"))
+                return false;
+
+            GraphServiceClient graphClient = GraphClient;
+
+            if (graphClient is null)
+                return false;
+
+            try
+            {
+                IUserRequest request = graphClient.Users[userName].Request();
+
+                // Load user data - note that external users need to be looked up by userPrincipalName
+                Microsoft.Graph.User user;
+                if (userName.Contains("#EXT#"))
+                    user = graphClient.Users.Request().Filter($"userPrincipalName eq '{userName}'").GetAsync().Result.FirstOrDefault();
+                else
+                    user = request.GetAsync().Result;
+
+                return user is not null;
+            }
+            catch (ServiceException ex)
+            {
+                if (ex.Error.Code == "Request_ResourceNotFound")
+                    return false;
+                else
+                    throw new Exception("Unable to query Azure", ex);
+            }
+            catch (Exception ex)
+            {
+                throw new Exception("Exception attempting to query Azure", ex);
+            }
+        }
```

Comment:
When I have multiple async function calls in a synchronous function, I usually like to wrap the async code up in a single async function and then await the result of that function. I feel it ends up cleaner and is easier to move the logic to an async function later if the need arises.

Furthermore, I wonder if `Task.Result` might have wrapped our `ServiceException` in an `AggregateException` so that our exception handler wouldn't have caught it? Awaiting the tasks should be safer from that perspective as well.

```suggestion
        private bool IsValidAzureADUserName(string userName)
        {
            async Task<bool> IsValidAsync()
            {
                if (userName.Contains("\\"))
                    return false;

                GraphServiceClient graphClient = GraphClient;

                if (graphClient is null)
                    return false;

                try
                {
                    IUserRequest request = graphClient.Users[userName].Request();

                    // Load user data - note that external users need to be looked up by userPrincipalName
                    Microsoft.Graph.User user;
                    if (!userName.Contains("#EXT#"))
                    {
                        user = await request.GetAsync();
                    }
                    else
                    {
                        IGraphServiceUsersCollectionPage page = await graphClient.Users
                            .Request()
                            .Filter($"userPrincipalName eq '{userName}'")
                            .GetAsync();

                        user = page.FirstOrDefault();
                    }

                    return user is not null;
                }
                catch (ServiceException ex)
                {
                    if (ex.Error.Code == "Request_ResourceNotFound")
                        return false;
                    else
                        throw new Exception("Unable to query Azure", ex);
                }
                catch (Exception ex)
                {
                    throw new Exception("Exception attempting to query Azure", ex);
                }
            }

            Task<bool> isValidTask = IsValidAsync();
            return isValidTask.GetAwaiter().GetResult();
        }
```